packages/js-utils/src/unique-by.ts

```diff
@@ -0,0 +1,10 @@
+const defaultCompare = ( a: any, b: any ): boolean => a === b;
+
+function uniqueBy( arr: any[], fn = defaultCompare ) {
+	return arr.reduce( ( acc, v ) => {
+		if ( ! acc.some( ( x: any ) => fn( v, x ) ) ) acc.push( v );
+		return acc;
+	}, [] );
+}
```

Comment:
I think we can do much better as far as types are concerned here.

```ts
function uniqueBy<T>( arr: T[], fn: ( left: T, right: T ) => boolean ): T[]
```

That should do...

```suggestion
function uniqueBy<T>( arr: T[], fn: ( a: T, b: T ): boolean = defaultCompare ): T[] {
	return arr.reduce( ( acc, v ) => {
		if ( ! acc.some( ( x: any ) => fn( v, x ) ) ) acc.push( v );
		return acc;
	}, [] as T[]);
}
```

Reply:
Yes, agreed. Although your suggestion shows me there's a syntax error:

<img width="565" alt="Screenshot 2021-02-22 at 15 35 37" src="https://user-images.githubusercontent.com/9202899/108722588-9e980480-7523-11eb-8480-43bad635764a.png">

should it be this?

```ts
function uniqueBy< T >( arr: T[], fn: ( a: T, b: T ) => boolean = defaultCompare ): T[] {
  // ...
}
```

Reply:
Yes, sorry!